### ISSUE TITLE
Modify image.get_rendition() to take advantage of prefetch_related

### ISSUE THREAD

**Issue by thenewguy** [wagtail/wagtail#947]
The [Picturefill polyfill](http://scottjehl.github.io/picturefill/) makes it possible to output responsive img tags using srcset with cross browser support.

image.get_rendition() does not take advantage of [prefetch_related](https://docs.djangoproject.com/en/1.7/ref/models/querysets/#prefetch-related) because of the way the rendition lookup is written.  (ref: [models.py#L172](https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailimages/models.py#L172))

Since a responsive img tag links to many renditions, rendering a lot of images greatly increases the number of database queries.  I.e. to output an image gallery with 30 images using 10 possible renditions per srcset suddenly takes 300 roundtrips to the database for renditions.

It would be beneficial to use 

```
            renditions = self.renditions.all()
            rendition = python_selection_func(renditions, filter, vary_key)
```

instead of 

```
            rendition = self.renditions.get(
                filter=filter,
                focal_point_key=vary_key,
            )
```

so that [prefetch_related](https://docs.djangoproject.com/en/1.7/ref/models/querysets/#prefetch-related) could be utilized.

**Comment by davecranwell** [comment]
@kaedroho we'll want to be using picture elements (or similar polyfills) in some projects coming soon, could you investigate and respond on this? You seem most eligible.

**Comment by kaedroho** [comment]
Thanks @thenewguy 

Another possibility on the API for this could be to create a new method on ``Image`` which takes a list of filters and returns a queryset of renditions. All of the optimisations can then be implemented in this method.

Also, we need to take into the account the fact that renditions may not exist and need to be generated on the fly. If none of the images that your gallery page needs are generated yet, that could cause 300 resizes to take place within the request. I'd be suprised if the request doesn't timeout before Wagtail has a chance to give a response.

**Comment by thenewguy** [comment]
A new method would be handy if you want to add one to the api.  It would also be good to refactor image and filter creation out of the current method at the same time for DRYness.  Otherwise both methods will need to duplicate that code.

I am currently generating the images in advance with celery.  Because of this I decide how many and what dimensions of renditions to offer the client automatically based on source dimensions alone.  Ideally the rendition url would be known in advance so the url could be returned without creating the rendition and renditions generated on demand.  But then there would also need to be a way to verify valid dimension choices based on source dimensions before creating and serving a rendition.  So generating them with celery in advance is good enough for now.

Sorry for any typos. I am currently traveling and sent this from my cell phone

**Comment by kaedroho** [comment]
Another idea could be to allow the ``{% image %}`` tag to return URLs to Wagtails external image rendering view instead of creating renditions. Generating these URLs is cheap and only requires the images id. It also won't block the initial request if the renditions don't exist.

This would mean that every image on the page will result in a request pointing at the Django backend, which could be solved by a combination of far-future cache headers and Varnish/a CDN (making only the first request to an image hit the backend).

**Comment by MattSegal** [comment]
Ran into this issue last week on a Wagtail (v1.13.1) site where we display ~60 company logos on the home page. I saw ~100 queries in the Django debug toolbar, which I cut down to ~20 when I monkey patched [`wagtail.wagtailimages.shortcuts.get_renditions_or_not_found`](https://github.com/wagtail/wagtail/blob/v1.13.1/wagtail/wagtailimages/shortcuts.py) with the following:

    def get_rendition_or_not_found_patch(image, filter):
        """
        This function is a monkey patch of 'get_rendition_or_not_found'. Here, we try to
        find the requested rendition in Python before calling the original function.
        
        This patch will allow you to use prefetch_related on renditions and grab all 
        your renditions in one query.
        """
        if type(filter) is str:
            filter = Filter(spec=filter)
        
        filter_spec = filter.spec
        focal_point_key = filter.get_cache_key(image)
        
        # This for loop assumes that you've prefetched your image's renditions
        rendition = None
        for rend in image.renditions.all():
            if rend.filter_spec == filter_spec and rend.focal_point_key == focal_point_key:
                rendition = rend
        
        if not rendition:
            rendition = get_rendition_or_not_found(image, filter)
        
        return rendition

    wagtail.wagtailimages.templatetags.wagtailimages_tags.get_rendition_or_not_found = get_rendition_or_not_found_patch

This exact approach isn't ideal because

- it assumes that you have prefetched your renditions
- if you haven't prefetched your renditions, it assumes that you only have a small number of renditions

Could we allow something _like_ this in the image tag though? Like an `{% image prefetched=True %}` kwarg or similar behaviour?

**Comment by jmfederico** [comment]
Just run into this!

Is any one working on a possible solution?

**Comment by zerolab** [comment]
There was some work in this area - #2461 /  #4856 introduce an `image_url` template tag ([usage docs](http://docs.wagtail.io/en/v2.5.1/advanced_topics/images/image_serve_view.html#generating-dynamic-image-urls-in-python)) that should help for this type of scenarios

**Comment by thaxy** [comment]
@zerolab not quite sure how ``{% image_url %}`` should help. If we are using the tag then we just move the queries visible in django debug toolbar (described by @MattSegal) to the view function ``wagtailimages_serve``.  Maybe I am missing the point but in my opinion, we are still hitting the database exactly the same amount of times as before. It only tackles the timeout issue described by @kaedroho.

**Comment by kaedroho** [comment]
@thaxy It'll defer image rendering to another request so we wouldn't need to run that query in the initial response, which should give a slight improvement to initial page load. You're right that this won't reduce the load on the server overall, but those image requests will be executed simultaneously (6 at a time) which might improve the overall page load speed for the user.

But the main speed advantage of handling images in separate requests is that you can easily cache them, so the image only needs to be rendered for the first user and all subsequent calls for that rendition would be served by a cache server or CDN.

**Comment by zerolab** [comment]
#8069 adds support for image rendition prefetching and will be in 3.0.
@thenewguy see https://docs.wagtail.org/en/latest/advanced_topics/images/renditions.html#prefetching-image-renditions - this should do what your initial post was asking.
There is also #7225 which adds a convenience method on `ImageQuerySet`

**Comment by zerolab** [comment]
Will mark this as closed as #8069 has been merged and is available since Wagtail 3
